src/components/menu/menu-item/menu-item.component.js

```diff
@@ -1,33 +1,211 @@
-import React from 'react';
+import React, {
+  useRef, useState, useEffect, useCallback
+} from 'react';
 import PropTypes from 'prop-types';
 import classNames from 'classnames';
 import StyledMenuItemWrapper from './menu-item.style';
 import { StyledSubmenu, StyledSubmenuItem, StyledSubmenuTitle } from '../submenu-block/submenu.style';
 import OptionHelper from '../../../utils/helpers/options-helper';
 import Link from '../../link';
+import Events from '../../../utils/helpers/events';
 
-const MenuItem = ({
+const MenuItem = React.forwardRef(({
   submenu,
   children,
   href,
   to,
   menuType,
   onClick,
   target,
-  submenuDirection,
+  submenuDirection = 'right',
   icon,
   selected,
-  routerLink
-}) => {
+  routerLink,
+  isFirstElement,
+  handleKeyDown,
+  isOpen,
+  setIsOpenSubmenu,
+  menuItemIndex
+}, ref) => {
+  const [didOpenByArrowUp, setOpenByArrowUp] = useState(false);
+  const [didOpenByMouseEnter, setOpenByMouseEnter] = useState(false);
+  const [didOpenByClick, setOpenByClick] = useState(false);
+  const submenuItemsRefs = submenu && useRef(React.Children.map(children, child => child.ref || React.createRef()));
+  const ifRef = () => !submenu && { ref };
+  const actualFocusedItemIndex = useRef();
+  const submenuWrapperRef = useRef();
+  const setFocusToSubmenuElement = useCallback((event, index) => {
+    if (event) {
+      event.preventDefault();
+    }
+
+    const findIndex = () => {
+      if (submenuItemsRefs.current[index] && submenuItemsRefs.current[index].current.tagName === 'DIV') {
+        if (Events.isDownKey(event)) {
+          return index + 1;
+        }
+        return index - 1;
+      }
+      return index;
+    };
+
+    actualFocusedItemIndex.current = findIndex();
+
+    const isFocusOnLastMenuItem = index === submenuItemsRefs.current.length;
+    const isFocusOnFirstMenuItem = index === -1;
+    if (isFocusOnLastMenuItem) {
+      actualFocusedItemIndex.current = 0;
+    } else if (isFocusOnFirstMenuItem) {
+      actualFocusedItemIndex.current = submenuItemsRefs.current.length - 1;
+    }
+
+    submenuItemsRefs.current[actualFocusedItemIndex.current].current.focus();
+  }, [submenuItemsRefs]);
+
+  const onCloseSubmenu = useCallback(() => {
+    setIsOpenSubmenu(null);
```

Comment:
The events in here are triggered by a document event listener, this means they are not batched.
https://github.com/facebook/react/issues/14259#issuecomment-439632622

This will cause three updates

- setIsOpenSubmenu
- setOpenByMouseEnter
- setOpenByArrowUp

this might cause some unexpected behaviour and will cause unnecessary re-renders.

We can avoid this by using a state object and updating the state in one go. You might want to use a state reducer to do this. https://reactjs.org/docs/hooks-reference.html#usereducer